app/app.go

```diff
@@ -547,13 +548,21 @@ func (app *App) InitChain(req abci.RequestInitChain) (res abci.ResponseInitChain
 	if req.ConsensusParams.Version.AppVersion == 0 {
 		panic("app version 0 is not accepted. Please set an app version in the genesis")
 	}
+	appVersion := req.ConsensusParams.Version.AppVersion
 
 	// mount the stores for the provided app version if it has not already been mounted
 	if app.AppVersion() == 0 && !app.IsSealed() {
-		app.mountKeysAndInit(req.ConsensusParams.Version.AppVersion)
+		app.mountKeysAndInit(appVersion)
 	}
 
-	return app.BaseApp.InitChain(req)
+	res = app.BaseApp.InitChain(req)
+
+	ctx := app.NewContext(false, tmproto.Header{})
+	if appVersion != v1 {
+		app.SetInitialAppVersionInConsensusParams(ctx, appVersion)
+		app.SetAppVersion(ctx, appVersion)
```

Comment:
unrelated comment:

setting up context for state access is a good idea in general, but when we have to initialize with an empty header it feels so broken since we are now relying on everything downstream of that context to not need or get data from that header. I don't like it lol.

Reply:
+1 don't like it. I couldn't see an alternative to save consensus params that doesn't use a context so not sure how to avoid it.

Reply:
Yeah I agree, it's not great.